### ISSUE TITLE
label with mlang strings shows error on manual activity completion

### ISSUE THREAD

**Issue by snack753** [iarenaza/moodle-filter_multilang2#29]
a text label with mlang strings shows the error: "SyntaxError JSON Parse error: Expected '}'" on manual activity completion.

**Comment by iarenaza** [comment]
@snack753 After digging in the code, I'm afraid that this is a big and deep issue in Moodle core. You can avoid that particular JSON error if you configure the `multilang2` plugin to filter "Content and Headings", instead of just "Content".

But that only masks the symptom, but doesn't fix the issue at all. The deeper issue here is that the label module sets (what Moodle code internally calls) the module "name" from the module "intro". The "intro" is the text you type in the label content, with your `multilang2` filter tags, etc. It then calls `format_string` on that "intro" text to get the "name" value. But `format_string` only filters the string if the filter is configured for "Content and Headings". If you configure the filter just for "Content", `format_string` doesn't do anything to the string. So the `multilang2` tags remain in the filtered string. But there is a second step in building the "name" value, that is trimming it to 50 characters max. This is the code that does that: https://github.com/moodle/moodle/blob/master/mod/label/lib.php#L37-L40

Those tags contain the `{` and `}` characters. So if the trimming keeps one of the opening `{` characters but removes the corresponding closing one (`}`) in the final "name" value, when the label "name" is sent as part of the JSON response it completely breaks the JSON syntax, and you get the above error.

By configuring the filter for "Content and Headings", you avoid that particular error. Because `format_string` does the filtering this time. And there are no `{` or `}` characters in your label "name" value (at least not from the filter tags themselves!). In fact, you can get the same JSON error without even using the `multilang2` filter. You only need to create a label with the text only containing a `{` character. 

But the whole thing is broken in other ways. If you do configure the filter to avoid the JSON error issue, the label "name" that gets stored in the database is the filtered value, in whatever language the user doing the label editing had configured at that time. So even if you have a label in two or more languages, the label name will only have one of the languages. And that name is what is shown as the tool-tip text when hovering the mouse over the "Mark as done" button. So it may happen that the label is currently being show in English, but the text of the tool-tip for the "Mark as done" button displays the text in Thai.

In summary, this is not a `multilang2` filter issue, but a big, deep problem with the core "label" module.

**Comment by iarenaza** [comment]
Just a small note to let you know that I digged into this issue. It turns out it's a bug in the Javascript AMD module for template processing. In particular, it assumes that any string that is going to be rendered must be a JSON object if it starts by an opening brace character[1] that is not immediately followed byt that same character. As in that case, it assumes in that case, that is a template parameter that needs to be interpolated[2].

This is the piece of code that implements the above (faulty) logic: https://github.com/moodle/moodle/blob/master/lib/amd/src/templates.js#L559-L562

[1] Or curly bracket, of however it's called in you English dialect, that is the `{` character)  
[2] Mustache templates use the `{{` and `}}` strings to delimit values to interpolate in the final string.

**Comment by snack753** [comment]
Changing the filter setting to "Content and Headings” helped to get rid of the error messages in my case.
Thanks for the feedback.

On 3 Dec 2022, at 6:49 PM, Iñaki Arenaza ***@***.***> wrote:


@snack753 <https://github.com/snack753> After digging in the code, I'm afraid that this is a big and deep issue in Moodle core. You can avoid that particular JSON error if you configure the multilang2 plugin to filter "Content and Headings", instead of just "Content".

But that only masks the symptom, but doesn't fix the issue at all. The deeper issue here is that the label module sets (what Moodle code internally calls) the module "name" from the module "intro". The "intro" is the text you type in the label content, with your multilang2 filter tags, etc. It then calls format_string on that "intro" text to get the "name" value. But format_string only filters the string if the filter is configured for "Content and Headings". If you configure the filter just for "Content", format_string doesn't do anything to the string. So the multilang2 tags remain in the filtered string. But there is a second step in building the "name" value, that is trimming it to 50 characters max. This is the code that does that: https://github.com/moodle/moodle/blob/master/mod/label/lib.php#L37-L40 <https://github.com/moodle/moodle/blob/master/mod/label/lib.php#L37-L40>
Those tags contain the { and } characters. So if the trimming keeps one of the opening { characters but removes the corresponding closing one (}) in the final "name" value, when the label "name" is sent as part of the JSON response it completely breaks the JSON syntax, and you get the above error.

By configuring the filter for "Content and Headings", you avoid that particular error. Because format_string does the filtering this time. And there are no { or } characters in your label "name" value (at least not from the filter tags themselves!). In fact, you can get the same JSON error without even using the multilang2 filter. You only need to create a label with the text only containing a { character.

But the whole thing is broken in other ways. If you do configure the filter to avoid the JSON error issue, the label "name" that gets stored in the database is the filtered value, in whatever language the user doing the label editing had configured at that time. So even if you have a label in two or more languages, the label name will only have one of the languages. And that name is what is shown as the tool-tip text when hovering the mouse over the "Mark as done" button. So it may happen that the label is currently being show in English, but the text of the tool-tip for the "Mark as done" button displays the text in Thai.

In summary, this is not a multilang2 filter issue, but a big, deep problem with the core "label" module.
